{"blogid": "119762449", "writerAge": "码龄8年", "writerBlogNum": "74", "writerCollect": "88", "writerComment": "27", "writerFan": "18", "writerGrade": "5级", "writerIntegral": "1993", "writerName": "一路向前ylc", "writerProfileAdress": "..\\..\\static\\writer_image\\profile_119762449.jpg", "writerRankTotal": "17356", "writerRankWeekly": "66196", "writerThumb": "37", "writerVisitNum": "174094", "blog_read_count": "22556", "blog_time": "于 2021-08-17 18:06:54 发布", "blog_title": "http 状态码499", "content": "<div class=\"article_content clearfix\" id=\"article_content\">\n<link href=\"../../static/bootstrap/css/csdnstyle.css\" rel=\"stylesheet\"/>\n<div class=\"htmledit_views\" id=\"content_views\">\n<h2>什么是499状态</h2>\n<p> nginx源码中对499状态码的定义如下：</p>\n<pre><code>\n/*\n * HTTP does not define the code for the case when a client closed\n * the connection while we are processing its request so we introduce\n * own code to log such situation when a client has closed the connection\n * before we even try to send the HTTP header to it\n */\n\n#define NGX_HTTP_CLIENT_CLOSED_REQUEST 499</code></pre>\n<p>总结：当客户端主动把连接断开时，HTTP 不为这种情形定义代码。同时我们处理它的请求时，我们引入了当一个客户端在我们尝试向其发送 HTTP 头之前关闭连接时，使用自己的代码（也就是 <a href=\"https://httpstatuses.com/499\"><code>499</code></a> 状态码）来记录这种情况。</p>\n<ul><li><span style=\"color:#494949;\"><code>499</code> 状态码不是</span> <a href=\"https://httpstatuses.com/\">HTTP 的标准代码</a></li><li><span style=\"color:#494949;\"><code>499</code> 状态码是 Nginx 自己定义，用来 <strong>记录</strong>（你没看错，就是记录一下） 服务端向客户端发送 HTTP 请求头之前，客户端已经关闭连接的一种情况</span></li><li><span style=\"color:#494949;\">最常见的场景就是 <code>timeout</code> 设置不合理，Nginx 把请求转发上游服务器，上游服务器慢吞吞的处理，客户端等不及了主动断开链接，Nginx 就负责记录了 <code>499</code></span></li></ul>\n<h2>什么情况Nginx记录499错误日志</h2>\n<p>这里我们使用 <a href=\"https://curl.haxx.se/\"><code>curl</code></a> 模拟请求一下，更多 <code>curl</code> 的骚操作请访问 <a href=\"https://www.ruanyifeng.com/blog/2019/09/curl-reference.html\">curl 的用法指南</a>。</p>\n<pre><code>for i in $(seq 1 10); do curl -m 2 \"http://api.example.test\"; done</code></pre>\n<p><img alt=\"\" height=\"436\" src=\"..\\..\\static\\image\\20210817175711857.png\" width=\"1200\"/></p>\n<p>解释： curl -m 2 \"http://api.example.test\"  2秒没有响应则断开</p>\n<pre><code>tail -f /var/log/nginx/apiexample.access.log\n\n172.19.0.1 - - [15/Nov/2019:06:32:19 +0000] \"GET / HTTP/1.1\" 499 0 \"-\" \"curl/7.67.0\"\n172.19.0.1 - - [15/Nov/2019:06:32:22 +0000] \"GET / HTTP/1.1\" 499 0 \"-\" \"curl/7.67.0\"\n172.19.0.1 - - [15/Nov/2019:06:32:24 +0000] \"GET / HTTP/1.1\" 499 0 \"-\" \"curl/7.67.0\"\n172.19.0.1 - - [15/Nov/2019:06:32:26 +0000] \"GET / HTTP/1.1\" 499 0 \"-\" \"curl/7.67.0\"\n172.19.0.1 - - [15/Nov/2019:06:32:28 +0000] \"GET / HTTP/1.1\" 499 0 \"-\" \"curl/7.67.0\"\n172.19.0.1 - - [15/Nov/2019:06:32:30 +0000] \"GET / HTTP/1.1\" 499 0 \"-\" \"curl/7.67.0\"\n172.19.0.1 - - [15/Nov/2019:06:32:32 +0000] \"GET / HTTP/1.1\" 499 0 \"-\" \"curl/7.67.0\"\n172.19.0.1 - - [15/Nov/2019:06:32:34 +0000] \"GET / HTTP/1.1\" 499 0 \"-\" \"curl/7.67.0\"\n172.19.0.1 - - [15/Nov/2019:06:32:36 +0000] \"GET / HTTP/1.1\" 499 0 \"-\" \"curl/7.67.0\"\n172.19.0.1 - - [15/Nov/2019:06:32:38 +0000] \"GET / HTTP/1.1\" 499 0 \"-\" \"curl/7.67.0\"\n</code></pre>\n<p>如上所见，使用 <code>Timeout</code> 很容易模拟出 <code>499</code> 这种情形。</p>\n<p>记录 <code>499</code> 的情形：</p>\n<ul><li><span style=\"color:#494949;\">如上所示，数据传输的最大允许时间超时的话，Curl 断开了请求，而 Web 服务器如 Nginx 还在处理的话，则 Nginx 会记录 <code>499</code></span></li><li><span style=\"color:#494949;\">如果 Nginx 作为反向代理时，Nginx 将请求分发至对应的处理服务器时，有两对超时参数的设置：<code>proxy_send_timeout</code> 和 <code>proxy_read_timeout</code>、<code>fastcgi_send_timeout</code> 和 <code>fastcgi_read_timeout</code>。两对参数默认的超时时间都是 <code>60s</code>。在 Nginx 出现 <code>499</code> 的情况下，可以结合请求断开前的耗时和这两对设定的时间进行对比，看一下是不是在 <code>proxy_pass</code> 或者 <code>fastcgi_pass</code> 处理时，设置的超时时间短了</span></li><li><span style=\"color:#494949;\">如果 PHP 操作超时。打开 <code>php.ini</code> 查看 <code>max_execution_time</code> 和 <code>max_input_time</code> 两个参数。两者分别是 <code>PHP</code> 程序执行的最长时间和表单提交的最长时间</span></li><li><span style=\"color:#494949;\">如果两次提交 POST 过快就会出现 <code>499</code> 的情况，Nginx 认为是不安全的连接，主动拒绝了客户端的连接</span></li><li><span style=\"color:#494949;\">相关负载均衡配置等</span></li></ul>\n<p>如何有效防止Nginx记录499错误</p>\n<p>综上所述，我们可以得出一个结论，HTTP 请求在指定的时间内没能拿到响应而关闭了连接，就会发生 Nginx 记录 <code>499</code> 错误的情况。这个涉及到两个重要的问题：<strong>时间问题</strong> 和 <strong>性能问题</strong>（性能问题太过宽泛就不提及了），所以解决这个问题也就从这两方面入手。</p>\n<p>当然还有配置 <a href=\"http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_ignore_client_abort\"><code>proxy_ignore_client_abort</code></a> 参数为 <code>on</code> 来解决的（让代理服务端不要主动关闭客户端的连接）。但是这样也有一定的风险，会拖垮服务器。发生这个错误，如果服务器 CPU 和 Memory 不算太高，一般是数据库和程序的问题，数据库处理较慢或者程序线程较低。结合情况调整，比如读写分离或者程序线程数调高。</p>\n<p>（如果后端是fastcgi，比如php-fpm，配置fastcgi_ignore_client_abort on; //如果后端是fastcgi，比如php-fpm ）</p>\n<p>文档中对 <code>proxy_ignore_client_abort</code> 参数的说明：</p>\n<blockquote>\n<p>Determines whether the connection with a proxied server should be closed when a client closes the connection without waiting for a response.</p>\n</blockquote>\n<p>翻译：当一个客户端关闭连接而不等待响应时，确定与代理服务器的连接是否应该关闭。</p>\n<p></p>\n<p></p>\n<p></p>\n</div>\n</div>", "first_tag": "PHP", "cpp": 0, "csharp": 0, "python": 0, "javascript": 0, "java": 0, "sql": 0, "php": 1, "time": "2021-08-17 18:06:54", "summary": "什么是状态源码中对状态码的定义如下：总结：当客户端主动把连接断开时，不为这种情形定义代码。同时我们处理它的请求时，我们引入了当一个客户端在我们尝试向其发送头之前关闭连接时，使用自己的代码也就是状态码来"}